src/main/scala/Hash.scala

```diff
@@ -1,19 +1,643 @@
+
+    color match {
+      case Black => 16 * roleIndex(role) + (count - 1)
+      case White => 16 * roleIndex(role) + (count - 1) + 16 * 5
+    }
+  }
+
+  private[chess] def hexToBytes(str: String): PositionHash = {
+    str.grouped(2).map(cc =>
+      (Character.digit(cc(0), 16) << 4 | Character.digit(cc(1), 16)).toByte
+    ).toArray take size
+  }
+
+  def apply(situation: Situation): PositionHash = {
+    val board = situation.board
+    val stm = situation.color
+
+    val actors = board.actors.values.map { a =>
+      actorMasks(actorIndex(a))
+    }
+
+    val turn = stm match {
+      case Black => Array(zeroMask)
+      case White => Array(whiteTurnMask)
+    }
+
+    val castling = (situation.history.castles.toList zip castlingMasks).map {
+      case (canCastle, castlingMask) =>
+        if (canCastle) castlingMask else zeroMask
+    }
+
+    // According to FIDE rules en-passant needs to be completely legal in order
+    // to make a difference with regards to threefold repetition.
+    val ep = situation.moves.values.flatten.find(_.enpassant) match {
+      case Some(move) =>
+        Array(enPassantMasks(move.dest.x - 1))
+      case None =>
+        Array(zeroMask)
+    }
+
+    // Hash in sepcial three-check data.
+    val checks = board.variant match {
+      case variant.ThreeCheck =>
+        val blackCount = Math.min(situation.history.checkCount.black, 3)
+        val whiteCount = Math.min(situation.history.checkCount.white, 3)
+
+        Array(
+          if (blackCount > 0) threeCheckMasks(blackCount - 1) else zeroMask,
+          if (whiteCount > 0) threeCheckMasks(whiteCount - 1 + 3) else zeroMask
+        )
+      case _ =>
+        Array(zeroMask)
+    }
 
-  private[chess] val size = 3
+    // Hash in special crazyhouse data.
+    val crazy = board.crazyData match {
+      case Some(data) =>
+        ((data.promoted.map {
+          pos => crazyPromotionMasks(posIndex(pos))
+        }) ++ (data.pockets.white.roles.groupBy(identity).map {
+          case (role, list) =>
+            crazyPocketMasks(crazyPocketIndex(role, White, list.size))
+        }) ++ (data.pockets.black.roles.groupBy(identity).map {
+          case (role, list) =>
+            crazyPocketMasks(crazyPocketIndex(role, Black, list.size))
+        })).toArray
+      case None =>
+        Array(zeroMask)
+    }
 
-  private def apply(str: String): PositionHash =
-    MessageDigest getInstance "MD5" digest (str getBytes "UTF-8") take size
+    val masks = actors ++ turn ++ castling ++ ep ++ checks ++ crazy
 
-  def apply(actors: Iterable[Actor], color: Color): PositionHash = apply {
-    actors.map { a =>
-      s"${a.piece.forsyth}${a.pos.key}"
-    }.mkString + color.letter
+    (masks.transpose.map { column =>
+      column.reduce((a, b) => (a ^ b).toByte)
+    }).toArray
   }
 
+  // The first 8 bytes of the following masks are compatible with the Polyglot
+  // opening book format. The second half is generated randomly.
+
```

Comment:
Indeed we don't. And in fact all the literals are parsed by hexToBytes at startup and all but 3 bytes per hash are dropped.

The thing I have in mind is the Opening explorer (ornicar/lila#1528). It could now simply instanciate `new Hash(/*size*/ 16)` and produce globally unique hashes for hundreds of millions of games.